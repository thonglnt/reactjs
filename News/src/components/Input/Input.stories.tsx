// lib
import { Story } from '@storybook/react'

// components
import Input, { InputProps } from './'

export default {
  title: 'Components/Input',
  component: Input,
}

const Template: Story<InputProps> = (args) => <Input {...args} />

export const InputDefault = Template.bind({})
InputDefault.args = {
  name: 'Email',
  placeholder: 'Enter name',
  type: 'text',
}

export const InputWithLabel = Template.bind({})
InputWithLabel.args = {
  label: 'Input name',
  name: 'Email',
  placeholder: 'Enter name',
  type: 'text',
}

export const InputError = Template.bind({})
InputError.args = {
  errorMessage: 'Invalid name',
  name: 'Email',
  placeholder: 'Enter name',
  type: 'text',
}
