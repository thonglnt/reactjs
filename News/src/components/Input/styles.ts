// lib
import styled from 'styled-components'

// themes
import { theme } from '@themes'

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
`

export const LabelStyled = styled.label`
  margin: 0 0 10px 3px;
`

export const InputStyled = styled.input`
  width: 260px;
  height: 50px;
  padding-left: 10px;
  border: 1px solid ${theme.Colors.lightSilver};
  border-radius: 5px;

  &:focus {
    outline-width: 0;
  }
`

export const ErrMsgStyled = styled.small`
  color: ${theme.Colors.cinnabar};
  padding-left: 3px;
`
