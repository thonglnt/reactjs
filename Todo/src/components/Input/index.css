.input-label {
  padding-right: 10px;
  font-size: 16px;
  font-weight: 500;
}

.input-wrapper {
  display: inline-grid;
}

.input {
  width: 200px;
  height: 25px;
  padding-left: 5px;
  border-radius: 3px;
  border: 1px solid var(--light-gray);
}

.input:focus {
  outline-width: 0;
  border: 1px solid var(--cadet-blue);
}

.err-msg {
  padding-top: 3px;
  color: var(--red);
  font-size: 14px;
  font-weight: 400;
}

.input-lg {
  width: 500px;
  height: 40px;
}

@media screen and (max-width: 992px) {
  .input-lg {
    width: 300px;
    height: 35px;
  }
}

@media screen and (max-width: 775px) {
  .input-lg {
    width: 200px;
    height: 30px;
  }
}
