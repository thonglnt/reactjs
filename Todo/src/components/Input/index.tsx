// Lib
import { InputHTMLAttributes, memo } from 'react';

export type InputProps = {
  id?: string;
  name: string;
  label?: string;
  type?: InputHTMLAttributes<HTMLElement>['type'];
  placeholder?: string;
  errorMessage?: string;
  defaultValue?: string;
  size?: 'input-lg';
  minLength?: number;
  isRequired?: boolean;
};

export const Input = ({
  id,
  name,
  label,
  type = 'text',
  placeholder,
  errorMessage,
  defaultValue,
  size,
  minLength,
  isRequired = false,
}: InputProps) => {
  return (
    <>
      <label htmlFor={id} className='input-label'>
        {label}
      </label>
      <div className='input-wrapper'>
        <input
          id={id}
          defaultValue={defaultValue}
          type={type}
          className={`input ${size}`}
          name={name}
          placeholder={placeholder}
          minLength={minLength}
          required={isRequired}
        />
        {errorMessage && <small className='err-msg'>{errorMessage}</small>}
      </div>
    </>
  );
};

export default memo(Input);
