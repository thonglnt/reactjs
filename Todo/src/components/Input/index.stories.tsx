// Lib
import { Story } from '@storybook/react';

// Components
import { Input, InputProps } from '@components/Input';

export default {
  title: 'Input',
  component: Input,
};

const Template: Story<InputProps> = (args) => <Input {...args} />;

export const Sample = Template.bind({});
Sample.args = {
  label: 'Input name',
};

export const Error = Template.bind({});
Error.args = {
  label: 'Input name',
  errorMessage: 'Input invalid',
};
