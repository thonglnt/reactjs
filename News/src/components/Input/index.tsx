// lib
import React, { InputHTMLAttributes, memo } from 'react'

// styles
import { ErrMsgStyled, InputStyled, LabelStyled, Wrapper } from './styles'

export type InputProps = {
  name: string
  type: InputHTMLAttributes<HTMLElement>['type']
  label?: string
  placeholder?: string
  errorMessage?: string
  defaultValue?: string
}

const Input = React.forwardRef<HTMLInputElement & HTMLTextAreaElement, InputProps>(
  (
    {
      name = '',
      label = '',
      type = 'text',
      placeholder = 'Please enter something.',
      errorMessage = '',
      defaultValue = '',
      ...rest
    }: InputProps,
    ref,
  ): JSX.Element => {
    return (
      <>
        <Wrapper>
          {label && <LabelStyled>{label}</LabelStyled>}
          <InputStyled
            defaultValue={defaultValue}
            ref={ref}
            placeholder={placeholder}
            type={type}
            name={name}
            {...rest}
          />
        </Wrapper>
        {errorMessage && <ErrMsgStyled>{errorMessage}</ErrMsgStyled>}
      </>
    )
  },
)

export default memo(Input)
