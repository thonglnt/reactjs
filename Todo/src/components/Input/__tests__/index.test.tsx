// Lib
import { render, screen } from '@testing-library/react';

// Components
import { Input, InputProps } from '@components/Input';

describe('Test Input component', () => {
  const initInput: InputProps = {
    id: '1',
    label: 'Name',
    name: 'name',
  };

  test('It should render correctly with initial Props', () => {
    const { container } = render(<Input {...initInput} />);

    expect(container).toMatchSnapshot();
  });

  test('It should match DOM Snapshot', () => {
    const { container } = render(
      <Input
        {...initInput}
        defaultValue='name'
        errorMessage='error'
        type='email'
        placeholder='Please enter name'
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test('It should render correctly with error message', () => {
    render(<Input errorMessage='error' {...initInput} />);

    expect(screen.getByText('error'));
  });
});
