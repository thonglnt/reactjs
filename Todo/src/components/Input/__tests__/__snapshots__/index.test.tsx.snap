// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Input component It should match DOM Snapshot 1`] = `
<div>
  <label
    class="input-label"
    for="1"
  >
    Name
  </label>
  <div
    class="input-wrapper"
  >
    <input
      class="input undefined"
      id="1"
      name="name"
      placeholder="Please enter name"
      type="email"
      value="name"
    />
    <small
      class="err-msg"
    >
      error
    </small>
  </div>
</div>
`;

exports[`Test Input component It should render correctly with initial Props 1`] = `
<div>
  <label
    class="input-label"
    for="1"
  >
    Name
  </label>
  <div
    class="input-wrapper"
  >
    <input
      class="input undefined"
      id="1"
      name="name"
      type="text"
      value=""
    />
  </div>
</div>
`;
